### PR TITLE
re-scale sample weights to 1 mean

### PR DESCRIPTION
**Pull request by alexander-rakhlin**
Small sample weights result in diminished gradients, and estimator never learns. This is a typical outcome of setting weights to 1/number_of_samples like sklearn AdaBoostClassifier does

### CODE REVIEW

**Comment by Laurae2** [comment]
What if the user intentionally wants small weights?

If there are small weights, the user needs to adjust the gradient/hessian thresholds using the appropriate provided parameters (like `min_sum_hessian`). There are cases where it is useful to have such small weights.

**Comment by alexander-rakhlin** [comment]
With small weights it never learns with scikit's AdaBoost, and it's very difficult to understand the reason.
It should give a warning as the minimum.

**Comment by Laurae2** [comment]
@alexander-rakhlin LightGBM is not scikit-learn/AdaBoost, and it is not difficult to understand the reason it does not learn as boosting is a mathematical process. If your weights are too small and your code is using a hard threshold based on gradient/hessian (supposing each value lower than 1 separately) larger than the sum of weights, then it is obvious it will not learn as the trees cannot grow.

This is a structural issue identical to the same as providing 150 observations for learning when the default parameters are not letting less than 100 observations in a leaf: https://github.com/Microsoft/LightGBM/issues/380

Mathematically the sum of weights do not matter itself, it changes only the impact of formulas you need to scale by yourself. You can learn with small weights as long the parameters are setup correctly. Try this in R and you will notice you will converge faster and better:

```r
library(lightgbm)
weights1 <- rep(1/100000, 6513) # sum of weights = 0.06513
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, weight = weights1)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
weights2 <- rep(1/100000, 1611)
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label, weight = weights2)
params <- list(objective = "regression", metric = "l2", device = "cpu", min_sum_hessian = 1e-5) # Adjusted accordingly
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   100,
                   valids,
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
```

**Comment by wxchan** [comment]
This feature, whether it will be added or not, should be implemented in c++ version.

**Comment by alexander-rakhlin** [comment]
I mean it's difficult to catch this error when LightGBM is used as estimator in another meta-model like AdaBoost where setting sample weights is automatic and non-transparent for a user.

**Comment by mxbi** [comment]
@alexander-rakhlin This sounds more like an issue with Adaboost than an issue with LightGBM - doesn't seem worth it to break the API in order to work better with Adaboost. Unless I'm misunderstanding, wouldn't it be better to propose the ability to set weights in Adaboost, instead of overriding this in LightGBM?

**Comment by Laurae2** [comment]
> @alexander-rakhlin With small weights it never learns with scikit's AdaBoost

> @alexander-rakhlin I mean it's difficult to catch this error when LightGBM is used as estimator in another meta-model like AdaBoost

The problem is AdaBoost, not LightGBM here.

In addition, there is no problem for learning under very small weights. If you take my previous example:

Small weights (sum = 0.06513):

```
[1]:	test's l2:3.28168e-09 
[2]:	test's l2:1.77035e-13 
```

Regular weights (sum = 6513):

```
[1]:	test's l2:0.0271235 
[2]:	test's l2:0.0147532 
```

**Comment by alexander-rakhlin** [comment]
@Laurae2 this might or might not be a problem of AdaBoost, it's surely a problem for an end user. No, LightGBM does not learn with small weights in the order of 1e-5. In more realistic task with ~100 features and 3 classes it outputs equal probabilities for every class and sample which is a sign of diminished gradient. Without weighting it's okay, just like with re-scaled to 1-mean weights.

**Comment by mxbi** [comment]
@alexander-rakhlin I think the choice should be up to the user, since everyone's use case is different - not everyone's datasets fall under your "more realistic task" umbrella, and LightGBM should let the user manually choose the parameters for their dataset, instead of forcing usually better parameters on them. I think a warning might be reasonable, but I don't think removing functionality just because another package uses the functionality incorrectly is a good idea.

**Comment by alexander-rakhlin** [comment]
@mxbi as far as I know, initializing sample weights to 1/n_features is typical for boosting algorithms that do sample-wise boosting (like AdaBoost, RankBoost). So I wouldn't say AdaBoost uses LightGBM incorrectly - it uses it according to practice developed in their scope.

**Comment by Laurae2** [comment]
@alexander-rakhlin the 1/n distribution initialization is an initialization the literature has criticized for being naive, meant to work in typical scenarios but not in edge cases like the one you are presenting.

There is a reason why AdaBoost falls in the class of "Weight Boosting": it is because it is based on weights. You need to use Cost-Generalized AdaBoost in that case, not AdaBoost.

In that specific case of weight initialization issue, using AdaBoost would be the same as using the wrong algorithm for the wrong task. This is not the fault of AdaBoost but also not of xgboost/LightGBM. This is the user trying to use the wrong tool for a specific problem (this is one of the reasons why Machine Learning is an iterative process, and not a monolithic process).

**Comment by alexander-rakhlin** [comment]
@Laurae2 nothing wrong with this combination: AdaBoost operates in sample space, LightGBM in feature space, they complement each other. With correct sample weighting AdaBoost improves LightGBM score.

**Comment by Laurae2** [comment]
> nothing wrong with this combination
> With small weights it never learns with scikit's AdaBoost

because you did not scale the hyperparameters correctly, LightGBM did nothing wrong.

Supposing sum of weights = 6513, if we switch to sum of weights = 0.06513 (x 1e-5) then

* if `min_sum_hessian = 10` then you will not learn because the hessian will never reach such gigantic value
* if `min_sum_hessian = 1e-4` then you will learn similarly to (sum of weight = 6513 and `min_sum_hessian = 10`)

> which is a sign of diminished gradient. 

This is not even the gradient issue you are talking about. If you don't believe it, run my code with `min_sum_hessian = 1e-4` and `min_sum_hessian=10` while the sum of weights is set to 0.06513. I do not know why you think this is a gradient issue, you can yourself check the values yourself by printing the sum of gradients and hessians using a custom objective. cf [debugging xgboost with custom objectives](https://www.kaggle.com/laurae2/allstate-claims-severity/fair-1/log)

```r
library(lightgbm)
weights1 <- rep(1/100000, 6513)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, weight = weights1)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
weights2 <- rep(1/100000, 1611)
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label, weight = weights2)
params <- list(objective = "regression", metric = "l2", device = "cpu", min_sum_hessian = 1e-4, nthread = 1)
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   50,
                   valids,
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
weight_loss <- as.numeric(model$record_evals$test$l2$eval)
plot(weight_loss)

params <- list(objective = "regression", metric = "l2", device = "cpu", min_sum_hessian = 10, nthread = 1)
model <- lgb.train(params,
                   dtrain,
                   50,
                   valids,
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
weight_loss <- as.numeric(model$record_evals$test$l2$eval)
plot(weight_loss)
```

`min_sum_hessian = 10` it is obvious you can't learn as the scaled sum of hessian will never reach such value:

image

`min_sum_hessian = 1e-4` it is mathematically obvious you can learn as you are effectively scaling the gradient and hessian by the ratio of old vs new weights:

image

**Comment by alexander-rakhlin** [comment]
> This is not even the gradient issue you are talking about.
You just confirmed my guess - this is indeed a gradient issue. Hessian is 2nd order derivative, that is gradient. I couldn't recreate this experiment because Scikit-learn API does not exposure min_sum_hessian, nor the documentation explains its connection with sample weights (it only says it can deal with over-fitting). But the issue of this topic holds - when you have no control over sample weights, they can be set automatically such that LightGBM never learns, and you can only guess what goes on, and can do nothing about that but patch LightGBM code.

**Comment by Laurae2** [comment]
@alexander-rakhlin I already said the cause in the first comment: https://github.com/Microsoft/LightGBM/pull/407#issuecomment-293634369 - and yes, LightGBM can learn with small weights only if you adjust parameters accordingly.

Now we are (again) getting at where is the fault: this is the fault of scikit-learn API, not the fault of LightGBM or AdaBoost. You need to fix scikit-learn API, not LightGBM API. LightGBM has full control over sample weights and hyperparameters, but in your case you are trying to use it by alienating this control. Therefore, the issue is not on LightGBM side.

In addition to your comment:

> nor the documentation explains its connection with sample weights (it only says it can deal with over-fitting)

You can't expect a documentation to be 100% complete, nor a documentation to be the equivalent of a full mathematics course on gradient descent with all the pitfalls and the potential issues of optimization. If you wish to fill the documentation, you are welcome to open PRs. Anyway, [xgboost](http://xgboost.readthedocs.io/en/latest//parameter.html) could already hint you it is a hessian issue and not a gradient issue if sum of hessian = 1:

> min_child_weight [default=1]
minimum sum of instance weight (hessian) needed in a child. If the tree partition step results in a leaf node with the sum of instance weight less than min_child_weight, then the building process will give up further partitioning. **In linear regression mode, this simply corresponds to minimum number of instances needed to be in each node.** The larger, the more conservative the algorithm will be.
range: [0,∞]

There is a reason we use issues: to raise issues, like #380 : now, those using LightGBM on small datasets are aware to change one or two critical parameters for learning.

To solve your problem with a long term solution (this must not be fixed here as it is a AdaBoost and scikit-learn API co-dependent issue only):

* Change AdaBoost weighting scheme (not realistic at all as it changes the algorithm - as you said, it is indeed the behavior of AdaBoost and is perfectly normal)
* Add a new initialization scheme in AdaBoost (realistic, but you would need empirical proofs for scikit-learn repository)
* Change scikit-learn API to allow to pass hyperparameters (more realistic, and probably way better welcome as I personally know people who would like scikit-learn API to be more extensible than it is currently now)

To solve your problem with a short term solution: fork this repository and add your code, and use your branch as it benefits your specific needs. This is what a lot of people do on other repositories when they have unique features they really need (but others won't need, as it causes a major issue).

**Comment by alexander-rakhlin** [comment]
> To solve your problem with a short term solution: fork this repository and add your code, and use your branch as it benefits your specific needs

Thank you for advice. I forked, and added, and used, and benefited my needs, and after that started this pull request  :)) As for patching scikit-learn, it was wrong idea. I tried it too, and it broke regression part of AdaBoost. As AdaBoost seems follows its classic it's better not to repair it

**Comment by guolinke** [comment]
According to the above comments. I don't think this PR is needed.
If you want to train with small weight data, you can set ```min_child_weight ``` to a small value.
